File: view/redux/types/notification.ts
export interface SMTPConfig {
  id: string;
  host: string;
  port: number;
  username: string;
  password: string;
  from_email: string;
  from_name: string;
  security: string;
  created_at: string;
  updated_at: string;
  is_active: boolean;
  user_id: string;
  organization_id: string;
}

export interface CreateSMTPConfigRequest {
  host: string;
  port: number;
  username: string;
  password: string;
  from_name: string;
  from_email: string;
  organization_id: string;
}

export interface UpdateSMTPConfigRequest {
  id: string;
  host?: string;
  port?: number;
  username?: string;
  password?: string;
  from_name?: string;
  from_email?: string;
  organization_id: string;
}

export interface PreferenceType {
  id: string;
  label: string;
  description: string;
  enabled: boolean;
}

export interface CategoryPreferences {
  category: 'activity' | 'security' | 'update';
  preferences: PreferenceType[];
}

export interface UpdatePreferenceRequest {
  category: 'activity' | 'security' | 'update';
  type:
    | 'password-changes'
    | 'security-alerts'
    | 'team-updates'
    | 'login-alerts'
    | 'product-updates'
    | 'newsletter'
    | 'marketing';
  enabled: boolean;
}

export interface GetPreferencesResponse {
  activity: PreferenceType[];
  security: PreferenceType[];
  update: PreferenceType[];
}

export interface PreferenceItem {
  id: string;
  preference_id: string;
  category: string;
  type: string;
  enabled: boolean;
}

export interface WebhookConfig {
  id: string;
  type: 'slack' | 'discord';
  webhook_url: string;
  webhook_secret?: string;
  channel_id: string;
  is_active: boolean;
  user_id: string;
  organization_id: string;
  created_at: string;
  updated_at: string;
}

export interface CreateWebhookConfigRequest {
  type: 'slack' | 'discord';
  webhook_url: string;
  webhook_secret?: string;
  channel_id: string;
}

export interface UpdateWebhookConfigRequest {
  type: 'slack' | 'discord';
  webhook_url?: string;
  webhook_secret?: string;
  channel_id?: string;
  is_active?: boolean;
}

export interface DeleteWebhookConfigRequest {
  type: 'slack' | 'discord';
}

export interface GetWebhookConfigRequest {
  type: 'slack' | 'discord';
}

export interface SMTPFormData {
  smtp_host: string;
  smtp_port: string;
  smtp_username: string;
  smtp_password: string;
  smtp_from_email: string;
  smtp_from_name: string;
}

File: api/internal/types/notification.go
package types

import (
	"time"

	"github.com/google/uuid"
	"github.com/uptrace/bun"
)

type NotificationPreferences struct {
	bun.BaseModel `bun:"table:notification_preferences,alias:np" swaggerignore:"true"`
	ID            uuid.UUID        `json:"id" bun:"id,pk,type:uuid"`
	UserID        uuid.UUID        `json:"user_id" bun:"user_id,notnull,type:uuid"`
	CreatedAt     time.Time        `json:"created_at" bun:"created_at,notnull,default:current_timestamp"`
	UpdatedAt     time.Time        `json:"updated_at" bun:"updated_at,notnull,default:current_timestamp"`
	DeletedAt     *time.Time       `json:"deleted_at,omitempty" bun:"deleted_at"`
	Items         []PreferenceItem `json:"items" bun:"rel:has-many,join:id=preference_id"`
}

type PreferenceItem struct {
	bun.BaseModel `bun:"table:preference_item,alias:npi" swaggerignore:"true"`
	ID            uuid.UUID `json:"id" bun:"id,pk,type:uuid"`
	PreferenceID  uuid.UUID `json:"preference_id" bun:"preference_id,notnull,type:uuid"`
	Category      string    `json:"category" bun:"category,notnull"`
	Type          string    `json:"type" bun:"type,notnull"`
	Enabled       bool      `json:"enabled" bun:"enabled,notnull"`
	CreatedAt     time.Time `json:"created_at" bun:"created_at,notnull,default:current_timestamp"`
	UpdatedAt     time.Time `json:"updated_at" bun:"updated_at,notnull,default:current_timestamp"`
}

type PreferenceCategory string

const (
	ActivityCategory PreferenceCategory = "activity"
	SecurityCategory PreferenceCategory = "security"
	UpdateCategory   PreferenceCategory = "update"
)

func CreateDefaultPreferenceItems(preferenceID uuid.UUID) []PreferenceItem {
	now := time.Now()
	items := []PreferenceItem{
		{
			ID:           uuid.New(),
			PreferenceID: preferenceID,
			Category:     string(ActivityCategory),
			Type:         "team-updates",
			Enabled:      true,
			CreatedAt:    now,
			UpdatedAt:    now,
		},
		{
			ID:           uuid.New(),
			PreferenceID: preferenceID,
			Category:     string(SecurityCategory),
			Type:         "login-alerts",
			Enabled:      true,
			CreatedAt:    now,
			UpdatedAt:    now,
		},
		{
			ID:           uuid.New(),
			PreferenceID: preferenceID,
			Category:     string(SecurityCategory),
			Type:         "password-changes",
			Enabled:      true,
			CreatedAt:    now,
			UpdatedAt:    now,
		},
		{
			ID:           uuid.New(),
			PreferenceID: preferenceID,
			Category:     string(SecurityCategory),
			Type:         "security-alerts",
			Enabled:      true,
			CreatedAt:    now,
			UpdatedAt:    now,
		},
		{
			ID:           uuid.New(),
			PreferenceID: preferenceID,
			Category:     string(UpdateCategory),
			Type:         "product-updates",
			Enabled:      true,
			CreatedAt:    now,
			UpdatedAt:    now,
		},
		{
			ID:           uuid.New(),
			PreferenceID: preferenceID,
			Category:     string(UpdateCategory),
			Type:         "newsletter",
			Enabled:      false,
			CreatedAt:    now,
			UpdatedAt:    now,
		},
		{
			ID:           uuid.New(),
			PreferenceID: preferenceID,
			Category:     string(UpdateCategory),
			Type:         "marketing",
			Enabled:      false,
			CreatedAt:    now,
			UpdatedAt:    now,
		},
	}

	return items
}

type SMTPConfigs struct {
	bun.BaseModel  `bun:"table:smtp_configs,alias:sc" swaggerignore:"true"`
	ID             uuid.UUID `json:"id" bun:"id,pk,type:uuid"`
	Host           string    `json:"host" bun:"host,notnull"`
	Port           int       `json:"port" bun:"port,notnull"`
	Username       string    `json:"username" bun:"username,notnull"`
	Password       string    `json:"-" bun:"password,notnull"`
	FromEmail      string    `json:"from_email" bun:"from_email,notnull"`
	FromName       string    `json:"from_name" bun:"from_name,notnull"`
	Security       string    `json:"security" bun:"security,notnull"`
	CreatedAt      time.Time `json:"created_at" bun:"created_at,notnull,default:current_timestamp"`
	UpdatedAt      time.Time `json:"updated_at" bun:"updated_at,notnull,default:current_timestamp"`
	IsActive       bool      `json:"is_active" bun:"is_active,notnull,default:false"`
	UserID         uuid.UUID `json:"user_id" bson:"user_id"`
	OrganizationID uuid.UUID `json:"organization_id" bun:"organization_id,notnull"`
}

type WebhookConfig struct {
	bun.BaseModel  `bun:"table:webhook_configs,alias:wc" swaggerignore:"true"`
	ID             uuid.UUID `json:"id" bun:"id,pk,type:uuid"`
	Type           string    `json:"type" bun:"type,notnull"`
	WebhookURL     string    `json:"webhook_url" bun:"webhook_url,notnull"`
	WebhookSecret  *string   `json:"webhook_secret,omitempty" bun:"webhook_secret"`
	ChannelID      string    `json:"channel_id" bun:"channel_id,notnull"`
	IsActive       bool      `json:"is_active" bun:"is_active,notnull,default:false"`
	UserID         uuid.UUID `json:"user_id" bun:"user_id,notnull,type:uuid"`
	OrganizationID uuid.UUID `json:"organization_id" bun:"organization_id,notnull,type:uuid"`
	CreatedAt      time.Time `json:"created_at" bun:"created_at,notnull,default:current_timestamp"`
	UpdatedAt      time.Time `json:"updated_at" bun:"updated_at,notnull,default:current_timestamp"`
}
